File: python/ames_houses/dslc_documentation/functions/prepare_ames_data.py
import pandas as pd
import numpy as np
import datetime


from functions.clean_ames_data import clean_ames_data
from functions.preprocess_ames_data import preprocess_ames_data



## load in the original data
ames_orig = pd.read_table("../data/AmesHousing.txt", sep="\t", header=0, 
                          na_values=["", "NA"], keep_default_na=False)


## The following code would define the training, validation, and test set equivalent to what was done in R:
## We won't run it because we are going to use the same versions of the training, validation, and test sets that
## were used in the book so that our results match the book as closely as possible
## These results will be slightly different because we can't match the random seed used in R

## filter to just the relevant portion of the data
# ames = ames_orig.copy()
# ames = ames.query("(`Sale Condition` == 'Normal') & (`MS Zoning` != ['A (agr)', 'C (all)', 'I (all)'])")
# ames = ames.drop(columns="Sale Condition")

## compute the date
# ames["date"] = pd.to_datetime(dict(year=ames["Yr Sold"], month=ames["Mo Sold"], day=1))
# split_date = ames["date"].quantile(0.6)
# split_date_month = int(split_date.month)
# split_date_year = int(split_date.year)

## this code would define the training, validation, and test set equivalent to what was done in R
# ames_train = ames.query("`Mo Sold` <= @split_date_month & `Yr Sold` <= @split_date_year")
## filter to houses not in training set
# ames_val = ames.query("~PID.isin(@ames_train.PID)")
## randomly select half of the houses for the validation set
# ames_val = ames_val.sample(round(len(ames_val.index)*0.5), random_state=3789)
## filter to houses not in training and validation sets for the test set
# ames_test = ames.query("~PID.isin(@ames_train.PID) & ~PID.isin(@ames_val.PID)")

## Since we want to use the same training, validation, test set that was used in the book, we will 
## instead load the versions of the training, validation, and test sets that were computed in R
## so that our results match the book as closely as possible
ames_train = pd.read_csv("../data/train_val_test/ames_train.csv",
                         na_values=["", "NA"], keep_default_na=False)
ames_val = pd.read_csv("../data/train_val_test/ames_val.csv",
                       na_values=["", "NA"], keep_default_na=False)
ames_test = pd.read_csv("../data/train_val_test/ames_test.csv",
                        na_values=["", "NA"], keep_default_na=False)

# clean the original data
ames_train_clean = clean_ames_data(ames_train)
ames_val_clean = clean_ames_data(ames_val)
ames_test_clean = clean_ames_data(ames_test)

ames_train_preprocessed = preprocess_ames_data(ames_train_clean)

# extract the neighborhoods included in the training data
# this is to ensure that the validation and test sets only include the 
# same neighborhoods as the training data
neighborhood_cols = list(ames_train_preprocessed.filter(regex="neighborhood").columns)
train_neighborhoods = [x.replace("neighborhood_", "") for x in neighborhood_cols]

# create preprocessed validation set
ames_val_preprocessed = preprocess_ames_data(
    ames_val_clean,
    column_selection=list(ames_train_preprocessed.columns),
    neighborhood_levels=train_neighborhoods
)

# create preprocessed test set
ames_test_preprocessed = preprocess_ames_data(
    ames_test_clean,
    column_selection=list(ames_train_preprocessed.columns),
    neighborhood_levels=train_neighborhoods
)

File: R/ames_houses/dslc_documentation/functions/prepareAmesData.R
# running the pre-processing and cleaning function for the Ames data
library(tidyverse)
library(lubridate)
library(janitor)

source("functions/cleanAmesData.R")
source("functions/preProcessAmesData.R")
# load in the original data
ames_orig <- read.table("../data/AmesHousing.txt",
                               sep = "\t", header = T)
# filter to just the relevant portion of the data
ames <- ames_orig |> 
  filter(Sale.Condition == "Normal",
         # remove agricultural, commercial and industrial
         !(MS.Zoning %in% c("A (agr)", "C (all)", "I (all)"))) |>
  dplyr::select(-Sale.Condition)

# split into training, validation, and testing
split_date <- ames_orig |>
  unite(date, Mo.Sold, Yr.Sold, 
        sep = "/", remove = FALSE) |>
  mutate(date = parse_date(date, "%m/%Y")) |>
  summarise(quantile(date, c(0.6), type = 1)) |>
  pull() |>
  ymd()
# Errata note: the following code is what is used in the book, but is incorrect. 
# define the training set
ames_train <- ames |>
  filter(Mo.Sold <= month(split_date), Yr.Sold <= year(split_date))
# This should be 
# ames_train <- ames |>
#    filter(Yr.Sold <= year(split_date) | (Mo.Sold <= month(split_date) & Yr.Sold == year(split_date)))


# define the validation set
set.seed(286734) # Do not change this seed once you have started using it!
ames_val <- ames |>
  # filter to the houses no in the training set
  filter(!(PID %in% ames_train$PID)) |>
  # take a random sample of half
  sample_frac(0.5)
ames_test <- ames |>
  # filter to the houses not in the training set or validation set
  filter(!(PID %in% c(ames_val$PID, ames_train$PID)))


# save the data so that we can use the same training, validation, and test 
# sets in python
write_csv(ames_train, "../data/train_val_test/ames_train.csv")
write_csv(ames_val, "../data/train_val_test/ames_val.csv")
write_csv(ames_test, "../data/train_val_test/ames_test.csv")


# clean the original data
ames_train_clean <- cleanAmesData(ames_train)
ames_val_clean <- cleanAmesData(ames_val)
ames_test_clean <- cleanAmesData(ames_test)




ames_train_preprocessed <- preProcessAmesData(ames_train_clean)

# extract the neighborhoods included in the training data
# this is to ensure that the validation and test sets only include the 
# same neighborhoods as the training data
train_neighborhoods <- ames_train_preprocessed |> 
  dplyr::select(contains("neighborhood_")) |>
  colnames() |>
  str_split("neighborhood_") |>
  map_chr(~.[2])

# create preprocessed validation set
ames_val_preprocessed <- 
  preProcessAmesData(ames_val_clean, 
                     column_selection = colnames(ames_train_preprocessed),
                     neighborhood_levels = train_neighborhoods,
                     keep_pid = TRUE)

# create preprocessed test set
ames_test_preprocessed <- 
  preProcessAmesData(ames_test_clean, 
                     column_selection = colnames(ames_train_preprocessed),
                     neighborhood_levels = train_neighborhoods)


